Source/qol/itemlabels.cpp

```diff
@@ -0,0 +1,170 @@
+{
+	altPressed = pressed;
+}
+
+bool IsGeneratingItemLabels()
+{
+	return generatingLabelsState == 1;
+}
+
+bool IsItemLabelHighlighted()
+{
+	return isLabelHighlighted;
+}
+
+void UpdateItemLabelOffsets(const CelOutputBuffer &out, BYTE *dst, int width)
+{
+	int xval = (dst - out.begin()) % out.pitch();
+	if (xval < drawMinX)
+		drawMinX = xval;
+	xval += width;
+	if (xval > drawMaxX)
+		drawMaxX = xval;
+}
+
+void GenerateItemLabelOffsets(const CelOutputBuffer &out)
+{
+	if (generatingLabelsState == 2)
+		return;
+	generatingLabelsState = 1;
+	int itemTypes = gbIsHellfire ? ITEMTYPES : 35;
+	for (int i = 0; i < itemTypes; i++) {
+		drawMinX = gnScreenWidth;
+		drawMaxX = 0;
+		CelClippedDrawTo(out, out.pitch() / 2 - 16, 351, itemanims[i], ItemAnimLs[i], 96);
+		labelCenterOffsets[i] = drawMinX - out.pitch() / 2 + (drawMaxX - drawMinX) / 2;
+	}
+	generatingLabelsState = 2;
+}
+
+void AddItemToLabelQueue(int x, int y, int id)
+{
+	if (altPressed == invertHighlightToggle)
+		return;
+	ItemStruct *it = &items[id];
+
+	char textOnGround[64];
+	if (it->_itype == ITYPE_GOLD)
+		sprintf(textOnGround, "%i gold", it->_ivalue);
+	else
+		sprintf(textOnGround, "%s", it->_iIdentified ? it->_iIName : it->_iName);
+
+	int nameWidth = GetTextWidth((char *)textOnGround);
+	x += labelCenterOffsets[ItemCAnimTbl[it->_iCurs]];
+	y -= TILE_HEIGHT;
+	if (!zoomflag) {
+		x <<= 1;
+		y <<= 1;
+	}
+	x -= nameWidth >> 1;
+	text_color clr = COL_WHITE;
+	if (it->_iMagical == ITEM_QUALITY_MAGIC)
+		clr = COL_BLUE;
+	if (it->_iMagical == ITEM_QUALITY_UNIQUE)
+		clr = COL_GOLD;
+	labelQ.push_back(itemLabelQueue({ id, it->_ix, it->_iy, x, y, nameWidth, 13, clr, textOnGround }));
+}
+
+void DrawItemNameLabels(const CelOutputBuffer &out)
+{
+	isLabelHighlighted = false;
+	const int borderX = 5;
+	for (unsigned int i = 0; i < labelQ.size(); ++i) {
+		std::unordered_set<int> backtrace;
+
+		bool canShow;
+		do {
+			canShow = true;
+			for (unsigned int j = 0; j < i; ++j) {
+				if (abs(labelQ[j].y - labelQ[i].y) < labelQ[i].height + 2) {
+					int newpos = labelQ[j].x;
+					if (labelQ[j].x >= labelQ[i].x && labelQ[j].x - labelQ[i].x < labelQ[i].width + borderX) {
+						newpos -= labelQ[i].width + borderX;
+						if (backtrace.find(newpos) != backtrace.end())
+							newpos = labelQ[j].x + labelQ[j].width + borderX;
+					} else if (labelQ[j].x < labelQ[i].x && labelQ[i].x - labelQ[j].x < labelQ[j].width + borderX) {
+						newpos += labelQ[j].width + borderX;
+						if (backtrace.find(newpos) != backtrace.end())
+							newpos = labelQ[j].x - labelQ[i].width - borderX;
+					} else
+						continue;
+					canShow = false;
+					labelQ[i].x = newpos;
+					backtrace.insert(newpos);
+				}
+			}
+		} while (!canShow);
+	}
+
+	for (const itemLabelQueue &t : labelQ) {
+
+		if (t.x < 0 || t.x >= gnScreenWidth || t.y < 0 || t.y >= gnScreenHeight)
+			continue;
+
+		if (MouseX >= t.x && MouseX <= t.x + t.width && MouseY >= t.y - t.height && MouseY <= t.y) {
```

Comment:
This and line 146 are great candidates for some sort of `IsPointInRect()`.

Reply:
`SDL_IsPointInRect` -- may also want to change `y` to be the top coordinate and have an `SDL_Rect` directly in the struct instead of `x, y, width, height`